### ISSUE TITLE
Never unpublish submissions [8,13,0.62]

### ISSUE THREAD

**Issue by stenington** [Iridescent-CM/technovation-app#2133]
The platform has code to "un-publish" a submission by removing the `published_at` dates if the submission doesn't appear to be complete. There's a big problem with that from a user point of view in that they receive an email saying they're published when it's initially set. Un-publishing without also sending an email alerting them is setting those teams up for failure.

I think one of two things would be better:
1. Never un-publish. Treat the `published_at` field as sacred. We can always run scripts to detect published submissions that don't appear to be complete if that's a potential issue.
2. Always send emails when `published_at` changes. Send one when it's set, and send another if it gets unset. That way the teams are alerted to potential problems and can fix them or notify us.

We should also probably record an activity on publish/unpublish. Our email logs right now only last 30 days, it would be nice to have a more permanent record of what's happening around important data events like this.

**Comment by hellafitz** [comment]
After discussion, Mike and I agree that option 1 is preferable. We'll need to work w/ the program team to have them triage the resulting list of incomplete submissions. The example here is a team that adds a member after completing submission and that new member either hasn't completed their account, or the new member is older and triggers new submission requirements (eg. a business plan).

**Comment by hellafitz** [comment]
Part of this ticket involves giving the program team a heads up

**Comment by hellafitz** [comment]
The mechanism for detecting published but incomplete submissions is in scope on this ticket.

**Comment by stenington** [comment]
The admin provides a `Complete?` column on Submissions. We could add more for `team.qualified?` and `published?`. Admins could then export reports and filter for cases where submissions are either marked as incomplete or where the team isn't qualified but the submission is published, which would indicate that something has gone wrong.

The `published_at` field basically controls which submissions get scored, so flagging these sorts of discrepancies is important.

Anywhere we used to un-publish could be updated to log noisily instead, since this most likely represents a future problem during judging.

We might also consider the flip-side of this issue, which is detecting completely eligible submissions that **are not** published. For that we might want to add a column to Submissions for `only_needs_to_submit?` and possibly add controls to allow admins to publish on behalf of users, since last year that was done periodically as a developer task. That's probably better as another ticket though.

**Comment by stenington** [comment]
There's some tricky semantics involved here. In the admin, the `Complete?` column tells you **both** if the submission has all required fields **and** it has been published. The `Complete` filter, however, is only filtering for submissions which have or have not been published. So it's possible to filter for "Complete" submissions and find ones that aren't "Complete?", which is very confusing at first.

I'm also not sure the program team talks about published or unpublished. I think they talk about submitted or unsubmitted instead. If that's true, my proposal would be to rename the "Complete" filter to "Submitted". I would also provide a column for "submitted?" instead of "published?" to match that language.

I'd also add a column for "Required fields" that reports "complete" or "incomplete" so that the two components of the "Complete?" column can be broken apart in reports.

**Comment by stenington** [comment]
And just to note it here, the app **does allow** users to continue editing submissions after publishing, so it would be possible for a student to publish but then remove some required component of their submission after doing so. So there are two competing use cases pulling opposite directions on the idea of automatically un-publishing:

1. a user can publish but then remove required pieces of their submissions, which suggests the platform should automatically un-publish
2. last year we had teams publish but then get unpublished by a bug or some action *not directly taken by them*, which suggests the platform should not automatically un-publish

I think the conservative approach is still what we decided above: never un-publish, but verify before judging that all published submissions are appropriate for judging, and either work with teams to fix issues or manually un-publish as needed.

**Comment by stenington** [comment]
I feel like my notes here are getting complicated, so here's another approach. [This documentation](https://docs.google.com/document/d/1BEHSSY9gSwKpUNJUZaVSwcixPC6HRPqNVy_33t9p7Ow/edit#) provides a lightweight description of the problem, and details the proposed steps an admin could take to help manage the issue this season. Comments in the document identify potential areas of work.

Ideally, the program team would agree to some version of this document that's possible to implement in the remaining time for this season, and we can use the approved document to identify the changes or additions to the platform necessary to make the documentation correct. @rsgonzal @hellafitz does that seem helpful?

**Comment by ophelieh** [comment]
Hi @stenington! I just reviewed the document and it all makes sense. I added a couple of comments.

**Comment by stenington** [comment]
QA: TBD

I owe a write-up on the starting point for QA here, but for now I'm going to summarize the actual work done.

In the admin, under **Submissions**...

**New columns**
* `Required fields`: reports "Complete" or "Incomplete" depending on whether the team has completed all required fields on their submission
* `Team qualified?`: reports "yes" or "no" depending on whether the team has students, and all students are onboarded
* `Submitted?`: reports "yes" or "no" depending on whether `published_at` has been set on the submission
* `Only needs to submit?`: reports "yes" if the team is qualified and all required fields are completed, but the team has not published, or "no" otherwise

**Changed filter**
* `Complete` is now called `Submitted` with values "Yes, submitted" or "No, unsubmitted"

**Changed behavior**
* the platform will no longer "unpublish" a submission, but will log a warning instead if a published app is detected to not have all the required fields

**Comment by stenington** [comment]
**Only need to submit scenario**

1. log in as admin
2. go to **Submissions**
3. search for **Submitted: No, unsubmitted**, add column **Only needs to submit?**
4. find a team on the list that only needs to submit (or if necessary create a new team and complete a submission *except* for clicking the submit button)
5. log in as a team member
6. submit the project (you should only need to click the button)
7. log back in as admin
8. search for **Submitted: Yes, submitted**
9. confirm the team you chose is there with 100% progress

**Submitted but incomplete scenario**

1. log in as the team from the previous steps
2. change something about your submission, e.g. delete the app name and save
3. log in as admin
4. go to **Submissions**
5. search for the team name, add columns **Required fields**, **Team qualified?**, **Submitted?**
6. confirm your team shows up as **Submitted: yes** and **Required fields: Incomplete** with less than 100% progress

**Comment by rsgonzal** [comment]
Went through the two scenarios and all works as expected!

Have not shared with the program team

**Comment by stenington** [comment]
@rsgonzal What's the status on this? Is someone on the program team on the hook for QA?

**Comment by rsgonzal** [comment]
Yes! My bad, I sent it to them a little late. Ophelie said she will take a look either today or Monday.

**Comment by devincdillon** [comment]
I have looked. Both scenarios work as expected.

**Comment by ophelieh** [comment]
Hi all,

Only need to submit scenario - Everything worked as expected

Submitted but incomplete scenario - Everything worked as expected.

Question 1: Why does the submission still show as submitted on the admin side when a required field is removed but on the team's submission page it shows as incomplete? I just want to make sure I completely understand. 
![image](https://user-images.githubusercontent.com/32206421/68981190-432b8280-07c8-11ea-90f5-f443e2a4c5aa.png)

Question 2: What does it mean when a team has completed all the required fields but only shows 75%? I went on the team's submission page and it looks like everything is complete - Team Gargle Blaster. 
![image](https://user-images.githubusercontent.com/32206421/68980675-c8159c80-07c6-11ea-8447-086ad5f71543.png)

On an undirect note, as we remove the ability to directly filter by complete/uncomplete submissions for everyone or just for our staff? 
![image](https://user-images.githubusercontent.com/32206421/68981081-edef7100-07c7-11ea-9028-393250793e43.png)

Thank you!!!

**Comment by rsgonzal** [comment]
Hey @ophelieh! I'll try to answer your questions and maybe Mike can give his input Monday

Question 1: Why does the submission still show as submitted on the admin side when a required field is removed but on the team's submission page it shows as incomplete? I just want to make sure I completely understand.
I think this is the problem with the platform, once it's been submitted it won't be unsubmitted - this is because there may be other platform bugs that would unsubmit projects when they shouldn't be. This was the best solution, I believe..

Question 2: What does it mean when a team has completed all the required fields but only shows 75%? I went on the team's submission page and it looks like everything is complete - Team Gargle Blaster.
You may be running into this because of the removal of the demo video! @stenington should this have been part of #

On an undirect note, are we removing the ability to directly filter by complete/uncomplete submissions for everyone or just for our staff? Screenshot below is the current platform.
That complete section has just been renamed "Submitted" - this describes the action better! Since it can be submitted and not complete :)

**Comment by ophelieh** [comment]
Ohhh ok. I think Mike had mentioned something about not being able to un-submit. Thank you for confirming! :) 

And haha I didn't even catch it was doing the same thing... Probably time to log off. 

Thank you and have a great weekend everyone! :)

**Comment by stenington** [comment]
@ophelieh Good catch on the 75% thing! It looks like the percentage complete was cached, and when I made some code changes the numbers didn't update correctly. I think I just fixed that in staging, "Gargle Blaster 3000" was recalculated as 100% anyway.

In production we shouldn't have to worry about it, because there won't yet be any submissions when we deploy.

Roxana's answers are correct for the other two: it's "safer" not to un-submit for now (we might be able to come up with a better solution later) and filtering by complete was renamed, but it's still there.
